sqlflow_models/gcn.py

```diff
@@ -0,0 +1,408 @@
+            type from_node_col: string.
+        :param to_node_col: Name for the column in database to be used as the to_node id of each edge.
+            type to_node_col: string.
+        """
+        super(GCN, self).__init__()
+
+        self.gc_layers = list()
+        self.gc_layers.append(GCNLayer(nhid, kernel_regularizer=tf.keras.regularizers.l2(5e-4), sparse_input=sparse_input))
+        for i in range(nlayer-1):
+            self.gc_layers.append(GCNLayer(nhid, kernel_regularizer=tf.keras.regularizers.l2(5e-4)))
+        self.gc_layers.append(GCNLayer(nclass))
+        self.keep_prob = 1 - dropout
+        self.dropout = tf.keras.layers.Dropout(dropout)
+        self.nshape = None
+        self.train_ratio = train_ratio
+        self.eval_ratio = eval_ratio
+        self.nlayer = nlayer
+        self.epochs = epochs
+        self.early_stopping = early_stopping
+        self.sparse_input = sparse_input
+        self.id_col = id_col
+        self.feature_col = feature_col
+        self.from_node_col = from_node_col
+        self.to_node_col = to_node_col
+        # try to load the result file
+        try:
+            with open('./results.pkl', 'rb') as f:
+                self.results = pickle.load(f)
+        except (FileNotFoundError, IOError):
+            self.results = None
+
+    def call(self, data):
+        x, adj = data
+        assert self.nshape is not None, "Should calculate the shape of input by preprocessing the data with model.preprocess(data)."
+        if self.sparse_input:
+            x = GCN.sparse_dropout(x, self.keep_prob, self.nshape)
+        else:
+            x = self.dropout(x)
+        for i in range(self.nlayer-1):
+            x = tf.keras.activations.relu(self.gc_layers[i](x, adj))
+            x = self.dropout(x)
+        x = self.gc_layers[-1](x, adj)
+
+        return tf.keras.activations.softmax(x)
+
+    def evaluate(self, data, y, sample_weight):
+        """Function to evaluate the model."""
+        return self.test(sample_weight, return_loss=True)
+
+    def predict(self, data):
+        """Function to predict labels with the model."""
+        x, adj = data
+        for i in range(self.nlayer-1):
+            x = tf.keras.activations.relu(self.gc_layers[i](x, adj))
+        x = self.gc_layers[-1](x, adj)
+        return tf.keras.activations.softmax(x)
+
+    @staticmethod
+    def sparse_dropout(x, keep_prob, noise_shape):
+        """Dropout for sparse tensors."""
+        random_tensor = keep_prob
+        random_tensor += tf.random_uniform(noise_shape)
+        dropout_mask = tf.cast(tf.floor(random_tensor), dtype=tf.bool)
+        pre_out = tf.sparse_retain(x, dropout_mask)
+        return pre_out * (1./keep_prob)
+
+    @staticmethod
+    def encode_onehot(labels):
+        classes = set(labels)
+        classes_dict = {c: np.identity(len(classes))[i, :] for i, c in enumerate(classes)}
+        labels_onehot = np.array(list(map(classes_dict.get, labels)), dtype=np.int32)
```

Comment:
Well, this seems to be not working, since when we calculating the `loss` and `accuracy`, we will use `tf.gather_nd` to get masked labels, and tensorflow does not support converting a `numpy array` with `map` value to a `tensor`.